File: install-page/propagate.py
"""Propagate an orbit forward with full uncertainty + event detection.

Reproducer for the "Propagate an Orbit" section of
https://empyrean-dynamics.com/install. The marked region is what the
install page shows; the rest is just the entry point that makes this
runnable as `python install-page/propagate.py`.
"""

from __future__ import annotations


def main() -> None:
    # spielberg:snippet:start
    import empyrean
    from empyrean import Epochs, TimeScale, UncertaintyMethod

    empyrean.download_data()
    empyrean.initialize()

    orbits = empyrean.query_sbdb(["Apophis"])
    epochs = Epochs.from_kwargs(mjd=[65000.0], scale=TimeScale.TDB)

    result = empyrean.propagate(
        orbits,
        epochs,
        uncertainty_method=UncertaintyMethod.SECOND_ORDER,
    )

    for i in range(len(result.events.summary)):
        ev = result.events.summary
        print(
            f"{ev.event_type.to_pylist()[i]:25s} "
            f"{ev.body.to_pylist()[i]:8s} "
            f"MJD {ev.epoch.to_numpy()[i]:.2f}"
        )

    # Typed event tables for specific categories
    print(f"Earth CA: {result.events.periapses.distance_km.to_numpy()[0]:.0f} km")

    # Save everything to a directory of Parquet files
    result.to_dir("apophis_2029/")
    # spielberg:snippet:end


if __name__ == "__main__":
    main()

File: install-page/propagate.rs
//! Propagate an orbit forward with full uncertainty + event detection.
//!
//! Reproducer for the "Propagate an Orbit" section of
//! <https://empyrean-dynamics.com/install>.

fn main() -> empyrean::Result<()> {
    // spielberg:snippet:start
    use empyrean::{
        Context, Epoch, EventConfig, Origin, PropagationConfig, UncertaintyMethod, query_sbdb,
        write_events_parquet,
    };

    // Loads / downloads kernels on first run.
    let ctx = Context::from_data_dir(None)?;

    let batch = query_sbdb(&["Apophis"], None)?;
    let orbit = batch.orbits.into_iter().next().expect("not found");

    let epochs = vec![Epoch::from_mjd_tdb(65000.0)];
    let prop = PropagationConfig {
        uncertainty_method: UncertaintyMethod::SecondOrder,
        events: EventConfig {
            close_approaches: true,
            ..Default::default()
        },
        ..Default::default()
    };

    let result = ctx.propagate(&[orbit], &epochs, &prop)?;

    for ev in result.events.iter() {
        let body = ev
            .body
            .map(|b| format!("{b:?}"))
            .unwrap_or_else(|| "\u{2014}".into());
        println!(
            "{:25}  {:8}  MJD {:.2}",
            ev.event_type,
            body,
            ev.epoch.mjd()
        );
    }

    // Typed event accessors per category
    if let Some(ev) = result
        .events
        .iter()
        .find(|e| e.event_type == "periapsis" && e.body == Some(Origin::Earth))
    {
        println!("Earth CA: {:.0} km", ev.distance_km);
    }

    // Save the detected events to a Parquet file
    write_events_parquet("apophis_2029_events.parquet", &result.events)?;
    // spielberg:snippet:end
    Ok(())
}
